sycl/doc/extensions/proposed/sycl_ext_intel_cache_controls.asciidoc

```diff
@@ -98,85 +98,210 @@ Below is a list of new compile-time constant properties supported with
 ```c++
 namespace sycl::ext::intel::experimental {
 
-enum class cache_level : /*unspecified*/ {
+enum class level : /*unspecified*/ {
```

Comment:
Isn't `level` too ambiguous. @rdeodhar, what is the motivation to drop "cache_" prefix?

Reply:
When used in a cache control the word "cache" is already in there. I felt it was redundant. For example, cache_control_write_through<cache_level::L1>

Reply:
But, I could use cache_level if that's the preference.

Reply:
Please leave it as `cache_level`.  This enum is currently being used by other extensions that don't have "cache" in the name, like in `prefetch_hint<cache_level::L1>`.

I think it would be a better idea to rename the `cache_control` part of things to avoid redundancy. If you collapse everything into one property, that would be more aligned with the `prefetch` proposal, would remove the need for multiple ways of spelling "cached", and would probably be simpler for developers to use:

For example, you could bundle everything into `read_hint` and `write_hint` properties with a single `cache_control` enum:

```c++
enum cache_control {
  cached,
  uncached,
  streaming,
  invalidate,
  const_cached,
  write_through,
  write_back
};
```

Which would be used as:
- `read_hint<cache_level::L1, cache_control::cached>`
- `read_hint<cache_level::L1, cache_control::uncached>`
- `read_hint<cache_level::L1, cache_control::streaming>`
- `read_hint<cache_level::L1, cache_control::invalidate>`
- `write_hint<cache_level::L1, cache_control::cached>`
- `write_hint<cache_level::L1, cache_control::uncached>`
- `write_hint<cache_level::L1, cache_control::streaming>`
- `write_hint<cache_level::L1, cache_control::write_through>`
- `write_hint<cache_level::L1, cache_control::write_back>`

In keeping with the prefetch proposal, you'd then have shorthands like `read_hint_L1_uncached`, which would be a lot shorter than the existing shorthands like `cache_control_read_uncached<cache_level::L1>` in the current proposal.

Combinations that don't make sense (like `read_hint<cache_level::L1, cache_control::write_back>`) could be specified to give an error.

Tagging @jbrodman and @gmlueck for comment here, too.

Reply:
Oh, I see.  I think @steffenlarsen had some ideas for how we could implement properties in a way that didn't enforce that restriction to allow properties like this.

I might be wrong about this, but I think the restriction is actually that the `property_key` has to be unique.  So, it may still be possible to use two values like `read_hint<cache_level::L1, cache_control::cached>` and `read_hint<cache_level::L2, cache_control::cached>` as long as they resolve to different keys.  Is that true, @steffenlarsen, @gmlueck?

---

While thinking about this I took another look at the extension specification, and I think there are some bugs/inconsistencies there.  It currently says:

> Note that a property specifies the cache behavior only for the indicated cache level. In order to specify the behavior for multiple cache levels, multiple properties should be specified.

...but all the properties take a parameter pack for multiple levels.

It also currently says:

> It is legal to specify several different cache control properties in the same `annotated_ptr`. However, all instances of `cache_control_read_type` must have different cache levels and all instances of `cache_control_write_type` must have difference cache levels.

...suggesting that we're already violating the rule that a named property can be specified only once.

I'll give some thought to whether there's an alternative way to specify this stuff.

Reply:
I agree with @Pennycook that we should leave the name as `cache_level`.

I just noticed that the [sycl_ext_oneapi_prefetch](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_prefetch.asciidoc) extension, which shares this `cache_level` enum, defines it in the `ext::oneapi::experimental` namespace whereas this extension defines it in the `ext::intel::experimental` namespace.  I think it's important to share the same enum between these extensions, so I'd suggest defining it in the "oneapi" namespace for this extension also.  It's OK for an "intel" extension to use a "oneapi" thing, but not vice versa.

With your recent changes, it seems like the enumerations `cache_control_read_type` and `cache_control_write_type` aren't used any more.  Can they be removed?

Reply:
The doc needs correction. It was written before I learned of the single property limitation. It looks like I didn't quite make it match the current implementation.

Reply:
Yes, I'll remove the enumerations cache_control_read_type and cache_control_write_type.
And OK, on making the cache_level enumeration match the prefetch extension.

Reply:
> I might be wrong about this, but I think the restriction is actually that the `property_key` has to be unique. So, it may still be possible to use two values like `read_hint<cache_level::L1, cache_control::cached>` and `read_hint<cache_level::L2, cache_control::cached>` as long as they resolve to different keys. Is that true,

Yes, I think it is true that the key must be unique.  However, the purpose of having separate "key" and "property" was to allow someone to query a property by its key.  Thus, there needs to be a 1:1 correspondence between a property and its key.  If `read_hint<cache_level::L1, cache_control::cached>` and `read_hint<cache_level::L2, cache_control::cached>` had different keys, how would you look them up?

Reply:
Can the naming be like `read_hint_cached<cache_level::L1>`?  This still has the redundant "cache", but it's shorter than what you have now.  Other properties like `read_hint_streaming<cache_level::L1>` don't have any redundancy.

For the two assertions, maybe they could be named like `read_assertion_invalidate_afterwards<cache_level::L1>` and `read_assertion_const_cached<cache_level::L1>`.

Reply:
> To figure out the cache controls which apply to a read, you have to check for the existence of all possible cache control properties separately

Do we think it's that important for a user to query the read hints?  I think it's important for the *implementation* to do this query, but the existing structure where the queries are like "Does this property contain any uncached read hints?" is probably sufficient for this.

I'm not opposed to using the `read_hint<cache_control...>` format, where the property value is a pack of `cache_control`.  However, I wonder if that would really be an improvement for applications.

Reply:
> Do we think it's that important for a user to query the read hints? I think it's important for the implementation to do this query, but the existing structure where the queries are like "Does this property contain any uncached read hints?" is probably sufficient for this.

I think it's worth discussing, at least.  The way that we've specified compile-time property lists and compile-time properties, they're not an implementation detail -- things like `get_property` are exposed as part of the public API, and so I think there's a good chance developers will build SYCL library classes and SYCL library functions whose behaviors can be modified by compile-time properties.

And even if end-users don't have to suffer this pain, we should acknowledge that we're making the implementation more complicated.  Once these cache control properties exist, they may be supported by more than just `annotated_ptr` (e.g., `group_load` and `group_store`, other functions we haven't invented yet).

> I'm not opposed to using the read_hint<cache_control...> format, where the property value is a pack of cache_control. However, I wonder if that would really be an improvement for applications.

I think it could be, but I'm not 100% sure either.

From a readability standpoint, I also think the grouping is clearer.  It's the same logic as before, but now it's the reader (not the implementation) that has to mentally collect and group together all the relevant properties.

```c++
// This property list modifies the read behavior of whatever it's attached to, and has some other properties
// All the information about read behavior is in one place
sycl::properties { A, read_hint<cache_control<cache_mode::cached, cache_level::L1, cache_level::L2>,
                                cache_control<cache_mode::uncached, cache_level::L3>>, B };

// This property list is the same as the above, but A and B can appear at arbitrary locations
// cached, uncached, streaming properties can be arbitrarily ordered too
sycl::properties { cache_control_read_cached<cache_level::L1>, A, cache_control_read_uncached<cache_level::L3>,
                   B, cache_control_read_cached<cache_level::L2> };
```

If we open things up to `read_hint` accepting more than just `cache_control` objects in its parameter pack, we also leave the door open to common shorthands like:

```c++
// Could be equivalent to "uncached" at all levels of cache
sycl::properties { read_hint<nontemporal> };
```

Reply:
The form below has the problem that property "cache_control_read_cached" appears twice.
```
// This property list is the same as the above, but A and B can appear at arbitrary locations
// cached, uncached, streaming properties can be arbitrarily ordered too
sycl::properties { cache_control_read_cached<cache_level::L1>, A, cache_control_read_uncached<cache_level::L3>,
                   B, cache_control_read_cached<cache_level::L2> };
```
This form is OK:
```
// This property list modifies the read behavior of whatever it's attached to, and has some other properties
// All the information about read behavior is in one place
sycl::properties { A, read_hint<cache_control<cache_mode::cached, cache_level::L1, cache_level::L2>,
                                cache_control<cache_mode::uncached, cache_level::L3>>, B };
```
However, I don't see how someone calling get_property could avoid iterating all the cache_mode and cache_level values to figure out what properties were specified.

Reply:
> The form below has the problem that property "cache_control_read_cached" appears twice.

Oops, thanks for fixing that.

> However, I don't see how someone calling get_property could avoid iterating all the cache_mode and cache_level values to figure out what properties were specified.

The difference that I'm imagining really impacts how and when that iteration has to happen.

With `get_property`, the developer has to call `get_property<T>` and test each `T` associated with a cache control.  I _think_ that the developer always has to check every cache control just to find out if it was used or not, because I don't see an obvious way to iterate through only the properties that exist in a property list.

If everything is grouped together into something like a `read_hint`, then there's only one thing that has to be passed to `get_property`.  The cache controls within a `read_hint` are a regular parameter pack, so the developer can iterate through only the cache controls that were specified as part of the property.

I think what I've said is true of our current implementation or property lists, anyway, because the template argument of a property list is not exposed to the user.  A property list stores a tuple of property values, but that's an implementation detail).  This may be a gap in the property list specification.

Reply:
OK, I've adopted this syntax:
 read_hint<cache_control<cache_mode::cached, cache_level::L1, cache_level::L2>,
                                cache_control<cache_mode::uncached, cache_level::L3>>, B };